File: src/infrastructure/import/csvReader.ts
/**
 * قارئ CSV بمعيار RFC 4180 + متطلبات spec/05:
 * BOM · UTF-8 · فواصل داخل الاقتباس · أسطر CRLF · اقتباس مزدوج داخل الحقل.
 *
 * لا مكتبة خارجية: القارئ 60 سطرًا، وإضافة مكتبة تخالف قاعدة «لا مكتبة بلا سبب».
 */

export interface CsvRow {
  /** رقم الصف في الملف كما يراه المستخدم (الترويسة = 1). */
  lineNumber: number
  cells: string[]
  /** النص الأصلي للصف — يُحفظ للمراجعة (spec/05). */
  raw: string
}

export interface CsvDocument {
  header: string[]
  rows: CsvRow[]
}

export class CsvError extends Error {}

/** حدود معلنة — spec/05: «ضع حدودًا للملف وعدد الصفوف والوقت مع رسالة واضحة». */
export const CSV_LIMITS = {
  maxBytes: 12 * 1024 * 1024, // 12 ميجابايت
  maxRows: 100_000,
  maxCellLength: 20_000,
} as const

export function stripBom(text: string): string {
  return text.charCodeAt(0) === 0xfeff ? text.slice(1) : text
}

export function parseCsv(input: string): CsvDocument {
  if (input.length > CSV_LIMITS.maxBytes) {
    throw new CsvError(
      `الملف أكبر من الحد المسموح (${Math.round(CSV_LIMITS.maxBytes / 1024 / 1024)} ميجابايت)`,
    )
  }

  const text = stripBom(input)
  const rows: CsvRow[] = []

  let cells: string[] = []
  let field = ''
  let inQuotes = false
  let lineNumber = 1
  let rawStart = 0

  const pushRow = (endIndex: number) => {
    cells.push(field)
    // تجاهل الأسطر الفارغة تمامًا
    if (!(cells.length === 1 && cells[0] === '')) {
      rows.push({ lineNumber, cells, raw: text.slice(rawStart, endIndex) })
      if (rows.length > CSV_LIMITS.maxRows) {
        throw new CsvError(`الملف فيه أكتر من ${CSV_LIMITS.maxRows} صف`)
      }
    }
    cells = []
    field = ''
    lineNumber++
    rawStart = endIndex + 1
  }

  for (let i = 0; i < text.length; i++) {
    const ch = text[i]

    if (inQuotes) {
      if (ch === '"') {
        if (text[i + 1] === '"') {
          field += '"' // اقتباس مهروب داخل حقل
          i++
        } else {
          inQuotes = false
        }
      } else {
        field += ch
      }
      continue
    }

    if (ch === '"') {
      inQuotes = true
    } else if (ch === ',') {
      cells.push(field)
      field = ''
    } else if (ch === '\r') {
      // CRLF: تُعالَج عند \n التالي
      if (text[i + 1] === '\n') continue
      pushRow(i)
    } else if (ch === '\n') {
      const end = text[i - 1] === '\r' ? i - 1 : i
      pushRow(end)
      rawStart = i + 1
    } else {
      field += ch
      if (field.length > CSV_LIMITS.maxCellLength) {
        throw new CsvError(`خانة في السطر ${lineNumber} أطول من الحد المسموح`)
      }
    }
  }

  if (inQuotes) {
    throw new CsvError(`اقتباس مفتوح ولم يُغلق — الملف ناقص أو تالف عند السطر ${lineNumber}`)
  }
  if (field !== '' || cells.length > 0) pushRow(text.length)

  if (rows.length === 0) throw new CsvError('الملف فاضي')

  const header = rows[0].cells.map((h) => h.trim())
  return { header, rows: rows.slice(1) }
}

File: native-app/core/src/commonMain/kotlin/app/masroufy/core/CsvReader.kt
package app.masroufy.core

/**
 * قارئ CSV — نقل `src/infrastructure/import/csvReader.ts` (RFC 4180 + spec/05: BOM، فواصل جوه الاقتباس، CRLF).
 * نفس الحدود ونفس رسايل الأخطاء ونفس «النص الأصلي» لكل صف.
 */
data class CsvRow(
    /** رقم الصف في الملف زي ما المستخدم شايفه (الترويسة = 1). */
    val lineNumber: Int,
    val cells: List<String>,
    /** النص الأصلي للصف — بيتحفظ للمراجعة. */
    val raw: String,
)

data class CsvDocument(val header: List<String>, val rows: List<CsvRow>)

class CsvError(message: String) : IllegalArgumentException(message)

object CsvLimits {
    const val MAX_BYTES = 12 * 1024 * 1024
    const val MAX_ROWS = 100_000
    const val MAX_CELL_LENGTH = 20_000
}

fun stripBom(text: String): String = if (text.isNotEmpty() && text[0].code == 0xFEFF) text.substring(1) else text

fun parseCsv(input: String): CsvDocument {
    if (input.length > CsvLimits.MAX_BYTES) throw CsvError("الملف أكبر من الحد المسموح (12 ميجابايت)")
    val text = stripBom(input)
    val rows = mutableListOf<CsvRow>()
    var cells = mutableListOf<String>()
    val field = StringBuilder()
    var inQuotes = false
    var lineNumber = 1
    var rawStart = 0

    fun pushRow(endIndex: Int) {
        cells.add(field.toString())
        // السطر الفاضي تمامًا بيتجاهل
        if (!(cells.size == 1 && cells[0].isEmpty())) {
            rows.add(CsvRow(lineNumber, cells, text.substring(rawStart, endIndex)))
            if (rows.size > CsvLimits.MAX_ROWS) throw CsvError("الملف فيه أكتر من ${CsvLimits.MAX_ROWS} صف")
        }
        cells = mutableListOf()
        field.clear()
        lineNumber++
        rawStart = endIndex + 1
    }

    var i = 0
    while (i < text.length) {
        val ch = text[i]
        if (inQuotes) {
            if (ch == '"') {
                if (i + 1 < text.length && text[i + 1] == '"') {
                    field.append('"') // اقتباس مهروب جوه الخانة
                    i++
                } else {
                    inQuotes = false
                }
            } else {
                field.append(ch)
            }
            i++
            continue
        }
        when (ch) {
            '"' -> inQuotes = true
            ',' -> {
                cells.add(field.toString())
                field.clear()
            }
            '\r' -> if (!(i + 1 < text.length && text[i + 1] == '\n')) pushRow(i) // CRLF بيتعالج عند \n
            '\n' -> {
                val end = if (i > 0 && text[i - 1] == '\r') i - 1 else i
                pushRow(end)
                rawStart = i + 1
            }
            else -> {
                field.append(ch)
                if (field.length > CsvLimits.MAX_CELL_LENGTH) throw CsvError("خانة في السطر $lineNumber أطول من الحد المسموح")
            }
        }
        i++
    }

    if (inQuotes) throw CsvError("اقتباس مفتوح ولم يُغلق — الملف ناقص أو تالف عند السطر $lineNumber")
    if (field.isNotEmpty() || cells.isNotEmpty()) pushRow(text.length)
    if (rows.isEmpty()) throw CsvError("الملف فاضي")

    return CsvDocument(rows[0].cells.map { JsText.trim(it) }, rows.drop(1))
}
